File: segundos 4 lenguajes/Java/Main.java
package Java;
public class Main {
    public static void main(String[] args) {

            /*
         * NOTA IMPORTANTE:
         * - Las clases abstractas no se pueden instanciar directamente.
         *   Por eso en este main solo se crean objetos de clases normales.
         *
         * - En Java no existen destructores como en C++.
         *   El manejo de memoria lo realiza automáticamente el Garbage Collector.
         */

        // 1. JugadorNormal
    
        JugadorNormal jn1 = new JugadorNormal();        // sin parámetros
        JugadorNormal jn2 = new JugadorNormal("Santiago", 15);  // con parámetros
        JugadorNormal jn3 = new JugadorNormal("Mafe");         // sobrecargado
        JugadorNormal jn4 = new JugadorNormal(jn2);          // copia

    
        // 2. JugadorPro

        JugadorPro jp1 = new JugadorPro();             // sin parámetros
        JugadorPro jp2 = new JugadorPro("Luis", 30);       // con parámetros
        JugadorPro jp3 = new JugadorPro(50);         // sobrecargado
        JugadorPro jp4 = new JugadorPro(jp2);          // copia

        // 3. JugadorVIP
        
        JugadorVIP jv1 = new JugadorVIP();
        JugadorVIP jv2 = new JugadorVIP("Sofia", 100);
        JugadorVIP jv3 = new JugadorVIP("Marta");
        JugadorVIP jv4 = new JugadorVIP(jv2);

        // 4. JugadorNovato
        
        JugadorNovato jno1 = new JugadorNovato();
        JugadorNovato jno2 = new JugadorNovato("Pepito", 2);
        JugadorNovato jno3 = new JugadorNovato(5);
        JugadorNovato jno4 = new JugadorNovato(jno2);

        // 5. JugadorExperto
    
        JugadorExperto je1 = new JugadorExperto();
        JugadorExperto je2 = new JugadorExperto("Raúl", 40);
        JugadorExperto je3 = new JugadorExperto("Andrea");
        JugadorExperto je4 = new JugadorExperto(je2);

        // MOSTRAR
        
        System.out.println("Instancias creadas de los jugadores:");
        System.out.println("JugadorNormal: " + jn2.nombre + " Nivel " + jn2.nivel);
        System.out.println("JugadorPro: " + jp2.nombre + " Nivel " + jp2.nivel);
        System.out.println("JugadorVIP: " + jv2.nombre + " Nivel " + jv2.nivel);
        System.out.println("JugadorNovato: " + jno2.nombre + " Nivel " + jno2.nivel);
        System.out.println("JugadorExperto: " + je2.nombre + " Nivel " + je2.nivel);
    }
}

File: segundos 4 lenguajes/js/Main.js
import { JugadorNormal } from "./JugadorNormal.js";
import { JugadorPro } from "./JugadorPro.js";
import { JugadorVIP } from "./JugadorVIP.js";
import { JugadorNovato } from "./JugadorNovato.js";
import { JugadorExperto } from "./JugadorExperto.js";

// NOTA IMPORTANTE:
// En JavaScript solo se pueden instanciar las CLASES NORMALES.
// Las “clases abstractas” definidas en este proyecto se usan solo
// como clases base y NO deben ser instanciadas directamente,
// ya que contienen métodos sin implementar.
// También, JavaScript no tiene destructores; la memoria se maneja
// automáticamente mediante el Garbage Collector.

console.log("INSTANCIAS DE CLASES NORMALES:");

const jn1 = new JugadorNormal();
const jn2 = new JugadorNormal("Santiago", 15);
const jn3 = new JugadorNormal("Mafe");
const jn4 = new JugadorNormal(null, null, jn2);

const jp1 = new JugadorPro();
const jp2 = new JugadorPro("Luis", 30);
const jp3 = new JugadorPro("Desconocido", 50);
const jp4 = new JugadorPro(null, null, jp2);

const jv1 = new JugadorVIP();
const jv2 = new JugadorVIP("Sofia", 100);
const jv3 = new JugadorVIP("Marta");
const jv4 = new JugadorVIP(null, null, jv2);

const jno1 = new JugadorNovato();
const jno2 = new JugadorNovato("Pepito", 2);
const jno3 = new JugadorNovato("Novato", 5);
const jno4 = new JugadorNovato(null, null, jno2);

const je1 = new JugadorExperto();
const je2 = new JugadorExperto("Raúl", 40);
const je3 = new JugadorExperto("Andrea");
const je4 = new JugadorExperto(null, null, je2);

console.log("JugadorNormal:", jn2);
console.log("JugadorPro:", jp2);
console.log("JugadorVIP:", jv2);
console.log("JugadorNovato:", jno2);
console.log("JugadorExperto:", je2);
